src/coreclr/vm/riscv64/singlestepper.cpp

```diff
@@ -0,0 +1,467 @@
+        (&pCtx->R0)[reg] = value;
+}
+
+inline uint64_t SignExtend(uint64_t value, unsigned int signbit)
+{
+    _ASSERTE(signbit < 64);
+
+    if (signbit == 63)
+      return value;
+
+    uint64_t sign = value & (1ull << signbit);
+
+    if (sign)
+        return value | (~0ull << signbit);
+    else
+        return value;
+}
+
+inline uint64_t BitExtract(uint64_t value, unsigned int highbit, unsigned int lowbit, bool signExtend = false)
+{
+    _ASSERTE((highbit < 64) && (lowbit < 64) && (highbit >= lowbit));
+    uint64_t extractedValue = (value >> lowbit) & ((1ull << ((highbit - lowbit) + 1)) - 1);
+
+    return signExtend ? SignExtend(extractedValue, highbit - lowbit) : extractedValue;
+}
+
+// Parse the instruction opcode. If the instruction reads or writes the PC it will be emulated by updating
+// the thread context appropriately and true will be returned. If the instruction is not one of those cases
+// (or it is but we faulted trying to read memory during the emulation) no state is updated and false is
+// returned instead.
+bool RiscV64SingleStepper::TryEmulate(T_CONTEXT *pCtx, uint32_t opcode, bool execute)
+{
+    LOG((LF_CORDB, LL_INFO100000, "RiscV64SingleStepper::TryEmulate(opcode=%x, execute=%s)\n", opcode, execute ? "true" : "false"));
+
+    // Track whether instruction emulation wrote a modified PC.
+    bool fRedirectedPc = false;
+
+    // Track whether we successfully emulated an instruction. If we did and we didn't modify the PC (e.g. a
+    // AUIPC instruction or a conditional branch not taken) then we'll need to explicitly set the PC to the next
+    // instruction (since our caller expects that whenever we return true m_pCtx->Pc holds the next
+    // instruction address).
+    bool fEmulated = false;
+
+    // TODO after "C" Standard Extension support implemented, add C.J, C.JAL, C.JR, C.JALR, C.BEQZ, C.BNEZ
+
+    if ((opcode & 0x7f) == 0x17) // AUIPC
+    {
+        fEmulated = true;
+        if (execute)
+        {
+            LOG((LF_CORDB, LL_INFO100000, "RiscV64SingleStepper::TryEmulate AUIPC\n"));
+
+            // U-immediate
+            uint64_t imm = opcode & ~0xfffull;
+            uint64_t Rd = BitExtract(opcode, 11, 7);
+
+            uint64_t value = m_originalPc + imm;
+            SetReg(pCtx, Rd, value);
+        }
+    }
+    else if ((opcode & 0x7f) == 0x6f) // JAL
+    {
+        fEmulated = true;
+        if (execute)
+        {
+            LOG((LF_CORDB, LL_INFO100000, "RiscV64SingleStepper::TryEmulate JAL\n"));
+
+            // J-immediate encodes a signed offset in multiples of 2 bytes
+            //      21       | 20                                               1 | 0
+            // inst[31]/sign | inst[19:12] | inst[20] | inst[30:25] | inst[24:21] | 0
+            uint64_t imm = SignExtend((BitExtract(opcode, 30, 21) << 1) | (BitExtract(opcode, 20, 20) << 11) |
+                                      (BitExtract(opcode, 19, 12) << 12) | (BitExtract(opcode, 31, 31) << 20), 21);
```

Comment:
I think sign bit (inst[31] << 20) in inst is checked in SignExtend with (1ull << 20) not 21. Could you check it is correct?

Reply:
Thanks! I miss count bits when write comments and just set wrong sign bit even didn't looked at real sign bit that I moved properly -_- , should be
```
            //      20       | 19                                               1 | 0
            // inst[31]/sign | inst[19:12] | inst[20] | inst[30:25] | inst[24:21] | 0
```
will fix here and other lines too.